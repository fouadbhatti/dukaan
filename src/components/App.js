import React from 'react'

// import VisibleItemList from '../containers/VisibleItemList'
// import AddItem from '../containers/AddItem'
// import PropTypes from 'prop-types'

import { Container, Col, Row, Button, Jumbotron } from 'reactstrap';

import Products from '../containers/Products';
import AppBar from '../containers/AppBar';
import ProductDetails from '../containers/ProductDetails';

import { Switch, Route, Link, Redirect } from 'react-router-dom';

const App = () => {
  return (
    <Container fluid={true}>
      <Route path='/' component={AppBar}/>

      <Row>
        <Col lg="12" xs="12">
          <Switch>
            <Route exact path='/' component={Products}/>
            <Route exact path='/products' component={Products}/>
            <Route exact path='/product/:id' component={ProductDetails}/>
            <Route component={NotFound}/>

          </Switch>
        </Col>
      </Row>

    </Container>
  )
}

const NotFound = () => {
  return (
    <Container fluid={true}>
      <Jumbotron>
        <h1 className="display-3">404 Not Found!</h1>
        <p className="lead">Page Not Found!</p>
        <hr className="my-2" />
        <p>You can redirect to Home.</p>
        <p className="lead">
          <Link to='/products'>Home</Link>
        </p>
      </Jumbotron>
    </Container>
  )
}


export default App
