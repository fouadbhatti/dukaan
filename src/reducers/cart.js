import ActionTypes from '../constants/ActionTypes';
import Storage from '../helpers/storage';

const defaultState = {
  items: [],
  count: 0
};

const cart = (state = defaultState, action) => {

  switch(action.type) {

    case ActionTypes.ADD_TO_CART:

      const items = state.items;
      const item = {
        ...action.item,
        quantity: action.quantity
      };

      const atIdx = items.findIndex(it => it.id === item.id);

      if(atIdx != -1) {
        items[atIdx].quantity += item.quantity;
      } else {
        items.push(item);
      }

      state.price = items.reduce((total, i) => ((i.price * i.quantity) + total), 0);
      state.count = items.reduce((count, i) => (i.quantity + count), 0);

      const newState = { ...state, items };

      Storage.write('cart', JSON.stringify(newState))
      .then(() => {
        console.log('Storage syncing success.');
      })
      .catch(() => {
        console.log('Storage syncing failed.');
      });

      return newState;

    case ActionTypes.CHECKOUT:

      Storage.purge('cart', JSON.stringify(newState))
      .finally(() => {
        location.href = 'https://www.google.com';
      });

      return state;

    default:
      return state
  }
}

export default cart;
