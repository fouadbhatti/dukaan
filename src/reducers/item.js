import ActionTypes from '../constants/ActionTypes';

const item = (state = {}, action) => {

  switch(action.type) {

    case ActionTypes.ADD_ITEM:

      return { ...action.item };

    default:
      return state
  }
}

export default item;
