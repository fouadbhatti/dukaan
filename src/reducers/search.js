import ActionTypes from '../constants/ActionTypes';
import { Suggestion } from '../models';
import { QueryTransformer } from '../helpers/utils';


let _allItems = [];

const search = (state = { suggestions: [], query: '' }, action) => {

  switch(action.type) {

    case ActionTypes.ALL_ITEMS:

      _allItems = [...action.items];

      return state;

    case ActionTypes.FETCH_SEARCH_SUGGESTIONS:

      const q = action.query.trim();

      if(q) {
        const matcher = new QueryTransformer(q);
        const result = _allItems.filter(p => p.name.match(matcher))
          .map(item => new Suggestion(item.name));

        return { ...state, query: action.query, suggestions: [...result] };
      } else {
        return { ...state, query: action.query, suggestions: [] };
      }

    default:
      return state;
  }
};

export default search;
