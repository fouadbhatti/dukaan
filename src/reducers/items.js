import ActionTypes from '../constants/ActionTypes';
import { QueryTransformer } from '../helpers/utils';
import item from './item';

let _items = [];
let _allItems = [];

const items = (state = [], action) => {
  switch(action.type) {

    case ActionTypes.ALL_ITEMS:
      _allItems = [...action.items];

      return state;

    case ActionTypes.APPLY_PRODUCTS_UPDATE:

      _items = action.items.map(i => item({}, { type: ActionTypes.ADD_ITEM, item: i }));

      return _items;

    case ActionTypes.APPLY_SEARCH_CRITERIA:

      const q = action.query.trim();

      if(q) {
        const matcher = new QueryTransformer(q);
        return _allItems.filter(p => p.name.match(matcher));
      } else {
        return _items;
      }

    default:
      return state;
  }
};

export default items;
