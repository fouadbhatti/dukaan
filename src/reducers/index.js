import { combineReducers } from 'redux';
import items from './items';
import search from './search';

const reducers = combineReducers({
  items,
  search
});

export default reducers;
