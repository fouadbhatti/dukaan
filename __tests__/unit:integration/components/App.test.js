import React from 'react';
import ReactDOM from 'react-dom';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import reducers from '../../../src/reducers';
import App from '../../../src/components/App';

describe('<App />', () => {
  it('should render without crashing', () => {
    const wrapper = shallow(<App />);
    const component = wrapper.dive();

    expect(component).toMatchSnapshot();
  });
});
