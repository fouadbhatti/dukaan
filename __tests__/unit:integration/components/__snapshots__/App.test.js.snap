// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Container
    fluid={true}
    tag="div"
  >
    <Route
      component={[Function]}
      path="/"
    />
    <Row
      tag="div"
    >
      <Col
        lg={12}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Switch>
          <Route
            component={[Function]}
            exact={true}
            path="/"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/about"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/products"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/product/:id"
          />
          <Route
            component={[Function]}
          />
        </Switch>
      </Col>
    </Row>
  </Container>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Route
          component={[Function]}
          path="/"
        />,
        <Row
          tag="div"
        >
          <Col
            lg={12}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Switch>
              <Route
                component={[Function]}
                exact={true}
                path="/"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/about"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/products"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/product/:id"
              />
              <Route
                component={[Function]}
              />
            </Switch>
          </Col>
        </Row>,
      ],
      "className": "container-fluid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "component": [Function],
          "path": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Col
            lg={12}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Switch>
              <Route
                component={[Function]}
                exact={true}
                path="/"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/about"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/products"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/product/:id"
              />
              <Route
                component={[Function]}
              />
            </Switch>
          </Col>,
          "tag": "div",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Switch>
              <Route
                component={[Function]}
                exact={true}
                path="/"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/about"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/products"
              />
              <Route
                component={[Function]}
                exact={true}
                path="/product/:id"
              />
              <Route
                component={[Function]}
              />
            </Switch>,
            "lg": 12,
            "tag": "div",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Route
                  component={[Function]}
                  exact={true}
                  path="/"
                />,
                <Route
                  component={[Function]}
                  exact={true}
                  path="/about"
                />,
                <Route
                  component={[Function]}
                  exact={true}
                  path="/products"
                />,
                <Route
                  component={[Function]}
                  exact={true}
                  path="/product/:id"
                />,
                <Route
                  component={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                  "exact": true,
                  "path": "/",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                  "exact": true,
                  "path": "/about",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                  "exact": true,
                  "path": "/products",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                  "exact": true,
                  "path": "/product/:id",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "component": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Route
            component={[Function]}
            path="/"
          />,
          <Row
            tag="div"
          >
            <Col
              lg={12}
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Switch>
                <Route
                  component={[Function]}
                  exact={true}
                  path="/"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/about"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/products"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/product/:id"
                />
                <Route
                  component={[Function]}
                />
              </Switch>
            </Col>
          </Row>,
        ],
        "className": "container-fluid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "path": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Col
              lg={12}
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Switch>
                <Route
                  component={[Function]}
                  exact={true}
                  path="/"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/about"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/products"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/product/:id"
                />
                <Route
                  component={[Function]}
                />
              </Switch>
            </Col>,
            "tag": "div",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Switch>
                <Route
                  component={[Function]}
                  exact={true}
                  path="/"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/about"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/products"
                />
                <Route
                  component={[Function]}
                  exact={true}
                  path="/product/:id"
                />
                <Route
                  component={[Function]}
                />
              </Switch>,
              "lg": 12,
              "tag": "div",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Route
                    component={[Function]}
                    exact={true}
                    path="/"
                  />,
                  <Route
                    component={[Function]}
                    exact={true}
                    path="/about"
                  />,
                  <Route
                    component={[Function]}
                    exact={true}
                    path="/products"
                  />,
                  <Route
                    component={[Function]}
                    exact={true}
                    path="/product/:id"
                  />,
                  <Route
                    component={[Function]}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/about",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/products",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/product/:id",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
